kernel/kernel-api/src/main/java/io/delta/kernel/internal/types/TypeWideningChecker.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright (2025) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.kernel.internal.types;
+
+import io.delta.kernel.types.*;
+
+/**
+ * Utility class for checking type widening compatibility according to the Delta protocol.
+ *
+ * <p>The Type Widening feature enables changing the type of a column or field in an existing Delta
+ * table to a wider type. This class implements the checks required by the protocol to ensure that
+ * only supported type changes are allowed.
+ *
+ * <p>Supported type changes as per protocol:
+ *
+ * <ul>
+ *   <li>Integer widening: {@code Byte -> Short -> Int -> Long}
+ *   <li>Floating-point widening: {@code Float -> Double}
+ *   <li>Floating-point widening: {@code Byte, Short or Int -> Double}
+ *   <li>Date widening: {@code Date -> Timestamp without timezone}
+ *   <li>Decimal widening: {@code Decimal(p, s) -> Decimal(p + k1, s + k2)} where {@code k1 >= k2 >=
+ *       0}
+ *   <li>Decimal widening: {@code Byte, Short or Int -> Decimal(10 + k1, k2)} where {@code k1 >= k2
+ *       >= 0}
+ *   <li>Decimal widening: {@code Long -> Decimal(20 + k1, k2)} where {@code k1 >= k2 >= 0}
+ * </ul>
+ */
+public class TypeWideningChecker {
+  private TypeWideningChecker() {}
+
+  /**
+   * Checks if a type change from sourceType to targetType is a supported widening operation.
```

Comment:
Can you also state if sourceType = targetType it will return false? I think this is what is implemented so lets make sure that is clear

Reply:
and there's a test for this `same type is allowd`